source/slang/slang-check-expr.cpp

```diff
@@ -1355,6 +1355,162 @@ namespace Slang
+                return CreateErrorExpr(memberRefExpr);
+            }
+            int elementIndex[2];
+            elementIndex[0] = 0;
+            elementIndex[1] = 0;
+            // Check for the preceding underscore
+            if (*coursor++ != '_') {
+                getSink()->diagnose(swizExpr, Diagnostics::invalidSwizzleExpr, swizzleText, baseElementType->toString());
+                return CreateErrorExpr(memberRefExpr);
+            }
+
+            // Check for one or zero indexing            
+            if (*coursor == 'm') {
+                // Can't mix one and zero indexing
+                if (oneIndexed == 1) {
+                    getSink()->diagnose(swizExpr, Diagnostics::invalidSwizzleExpr, swizzleText, baseElementType->toString());
+                    return CreateErrorExpr(memberRefExpr);
+                }
+                oneIndexed = 0;
+                // Increment the index since we saw 'm'
+                coursor++;
+            }
+            else {
+                // Can't mix one and zero indexing
+                if (oneIndexed == 0) {
+                    getSink()->diagnose(swizExpr, Diagnostics::invalidSwizzleExpr, swizzleText, baseElementType->toString());
+                    return CreateErrorExpr(memberRefExpr);
+                }
+                oneIndexed = 1;
+            }
+
+            // Check for the ij components
+            for (Index j = 0; j < 2; j++) {
+                auto ch = *coursor++;
+                int subIndex;
+
+                // Manually enumerating cases to throw out 5+ and letters
+                switch (ch)
+                {
+                case '0': subIndex = 0; break;
+                case '1': subIndex = 1; break;
+                case '2': subIndex = 2; break;
+                case '3': subIndex = 3; break;
+                case '4': subIndex = 4; break;
+                default:
+                    // An invalid character in the swizzle is an error
+                    getSink()->diagnose(swizExpr, Diagnostics::invalidSwizzleExpr, swizzleText, baseElementType->toString());
+                    return CreateErrorExpr(memberRefExpr);
+                }
+
+                // Check the limit for either the row or column, depending on the step
+                IntegerLiteralValue elementLimit;
+                if (j == 0) {
+                    elementLimit = baseElementRowCount;
+                    elementIndex[0] = subIndex;
+                }
+                else {
+                    elementLimit = baseElementColCount;
+                    elementIndex[1] = subIndex;
+                }
+                // Make sure the index is in range for the source type
+                // Account for off-by-one and reject 0 if oneIndexed
+                if (subIndex >= elementLimit + oneIndexed || subIndex < oneIndexed)
+                {
+                    getSink()->diagnose(swizExpr, Diagnostics::invalidSwizzleExpr, swizzleText, baseElementType->toString());
```

Comment:
This might be improved with an error reporting the value is out of range and giving the value, and the range.

Reply:
I decided to skip on this requested change to match the HLSL style of just stating that the swizzle is invalid (and avoiding complicating the code with several error message types).  I can add a couple of error message types if you still disagree and think it's worth the overhead.